sdk/core/azure-core/inc/azure/core/http/policy.hpp

```diff
@@ -321,55 +321,25 @@ namespace Azure { namespace Core { namespace Http {
     void operator=(BearerTokenAuthenticationPolicy const&) = delete;
 
   public:
-    /**
-     * @brief Construct a Bearer Token authentication policy with single authentication scope.
-     *
-     * @param credential A #TokenCredential to use with this policy.
-     * @param scope Authentication scope.
-     */
-    explicit BearerTokenAuthenticationPolicy(
-        std::shared_ptr<TokenCredential const> credential,
-        std::string scope)
-        : m_credential(std::move(credential))
-    {
-      m_scopes.emplace_back(std::move(scope));
-    }
-
-    /**
-     * @brief Construct a Bearer Token authentication policy with multiple authentication scopes.
-     *
-     * @param credential A #TokenCredential to use with this policy.
-     * @param scopes A vector of authentication scopes.
-     */
-    explicit BearerTokenAuthenticationPolicy(
-        std::shared_ptr<TokenCredential const> credential,
-        std::vector<std::string> scopes)
-        : m_credential(std::move(credential)), m_scopes(std::move(scopes))
-    {
-    }
-
     /**
      * @brief Construct a Bearer Token authentication policy with multiple authentication scopes.
      *
-     * @tparam A type of scopes sequence iterator.
-     *
      * @param credential A #TokenCredential to use with this policy.
-     * @param scopesBegin An iterator pointing to begin of the sequence of scopes to use.
-     * @param scopesEnd An iterator pointing to an element after the last element in sequence of
-     * scopes to use.
+     * @param getTokenOptions #GetTokenOptions.
      */
-    template <typename ScopesIterator>
     explicit BearerTokenAuthenticationPolicy(
         std::shared_ptr<TokenCredential const> credential,
-        ScopesIterator const& scopesBegin,
-        ScopesIterator const& scopesEnd)
-        : m_credential(std::move(credential)), m_scopes(scopesBegin, scopesEnd)
+        GetTokenOptions const& getTokenOptions)
+        : m_credential(std::move(credential))
     {
+      m_getTokenOptions.Scopes = getTokenOptions.Scopes;
+      m_getTokenOptions.TransportPolicy.reset(
+          static_cast<Http::TransportPolicy*>(getTokenOptions.TransportPolicy->Clone().release()));
```

Comment:
Is this static cast safe? What exactly are we trying to do here with the clone/release of the transport policy passed in? Could this be a `std::move` instead?

Reply:
It is safe. We have to have it. `Clone()` returns `unique_ptr<HttpPolicy>`. The struct needs `unique_ptr<TransportPolicy>`. But `TransportPolicy->Clone()` still returns `unique_ptr<HttpPolicy>`. `TransportPolicy` derives from `HttpPolicy`. `static_cast` would not compile if they are unrelated. There is no cast from `unique_ptr<Base>` to `unique_ptr<Derived>`. So, you have to do `unique_ptr<Base>.release()` to get `Base*` (and `release()` tells `unique_ptr<Base>` to release control of it - it returns the original ptr, puts null on its place, so it would not dealloc the ptr in `~unique_ptr<Base>()` destructor). `reset()` takes ownership of `T*` (in this case, `Derived*` which we get by casting `static_cast<Derived*>(basePtr)`). As a result, we end up with the `unique_ptr<Derived>`.
Even if we didn't have to do the cast thing, we cannot use `move()`, because that would require to take the `options` by value, i.e. as a copy, but `unique_ptr`s are non-copyable, because they have exclusive ownership, and you can't (reasonably) have two `unique_ptr`s pointing at the same object.